ndcube/extra_coords/extra_coords.py

```diff
@@ -0,0 +1,321 @@
+import abc
+import copy
+from typing import Any, Tuple, Union, Iterable
+from numbers import Integral
+from functools import reduce
+from collections import defaultdict
+
+import astropy.units as u
+from astropy.modeling import models
+from astropy.wcs.wcsapi import BaseHighLevelWCS, BaseLowLevelWCS
+from astropy.wcs.wcsapi.sliced_low_level_wcs import sanitize_slices
+
+from .lookup_table_coord import LookupTableCoord
+
+__all__ = ['ExtraCoords']
+
+
+class ExtraCoordsABC(abc.ABC):
+    """
+    A representation of additional world coordinates associated with pixel axes.
+
+    ExtraCoords can be initialised by either specifying a
+    `~astropy.wcs.wcsapi.LowLevelWCS` object and a ``mapping``, or it can be
+    built up by specifying one or more lookup tables.
+
+    Parameters
+    ----------
+    wcs
+        The WCS specifying the extra coordinates.
+    mapping
+       The mapping between the array dimensions and pixel dimensions in the
+       extra coords object. This is an iterable of ``(array_dimension, pixel_dimension)`` pairs
+       of length equal to the number of pixel dimensions in the extra coords.
+
+    """
+
+    @abc.abstractmethod
+    def __init__(self,
+                 *,
+                 wcs: BaseLowLevelWCS = None,
```

Comment:
Should this also accept a `HighLevelWCS`?

Reply:
Arguably it could, but I am not sure it's needed. (It would just strip it down to low level anyway).

Reply:
:+1: